File: src/main/scala/resonantinduction/archaic/blocks/ItemImprint.java
package resonantinduction.archaic.blocks;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

import net.minecraft.entity.Entity;
import net.minecraft.entity.EntityList;
import net.minecraft.entity.EntityLivingBase;
import net.minecraft.entity.IProjectile;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.item.Item;
import net.minecraft.item.ItemStack;
import net.minecraft.nbt.NBTTagCompound;
import net.minecraft.nbt.NBTTagList;
import resonant.lib.utility.LanguageUtility;
import resonant.lib.utility.nbt.NBTUtility;
import resonantinduction.core.Reference;
import resonantinduction.core.ResonantTab;
import resonantinduction.core.Settings;

public class ItemImprint extends Item
{
	public ItemImprint()
	{
		super();
		this.setUnlocalizedName(Reference.prefix() + "imprint");
		this.setTextureName(Reference.prefix() + "imprint");
		this.setCreativeTab(ResonantTab.tab());
		this.setHasSubtypes(true);
		this.setMaxStackSize(1);
	}

	@Override
	public boolean onLeftClickEntity(ItemStack stack, EntityPlayer player, Entity entity)
	{
		if (entity != null && !(entity instanceof IProjectile) && !(entity instanceof EntityPlayer))
		{
			String stringName = EntityList.getEntityString(entity);
			// TODO Add to filter
			// player.sendChatToPlayer("Target: " + stringName);
			return true;
		}
		return false;
	}

	public boolean itemInteractionForEntity(ItemStack par1ItemStack, EntityLivingBase par2EntityLiving)
	{
		return false;
	}

	@Override
	public void addInformation(ItemStack itemStack, EntityPlayer par2EntityPlayer, List list, boolean par4)
	{
		Set<ItemStack> filterItems = getFilters(itemStack);

		if (filterItems.size() > 0)
		{
			for (ItemStack filterItem : filterItems)
			{
				list.add(filterItem.getDisplayName());
			}
		}
		else
		{
			list.add(LanguageUtility.getLocal("tooltip.noImprint"));
		}
	}

	/** Saves the list of items to filter out inside. */
	public static void setFilters(ItemStack itemStack, Set<ItemStack> filterStacks)
	{
		if (itemStack.getTagCompound() == null)
		{
			itemStack.setTagCompound(new NBTTagCompound());
		}

		NBTTagList nbt = new NBTTagList();

		for (ItemStack filterStack : filterStacks)
		{
			NBTTagCompound newCompound = new NBTTagCompound();
			filterStack.writeToNBT(newCompound);
			nbt.appendTag(newCompound);
		}

		itemStack.getTagCompound().setTag("Items", nbt);
	}

	public static boolean isFiltering(ItemStack filter, ItemStack itemStack)
	{
		if (filter != null && itemStack != null)
		{
			Set<ItemStack> checkStacks = getFilters(filter);

			if (checkStacks != null)
			{
				for (ItemStack stack : checkStacks)
				{
					if (stack.isItemEqual(itemStack))
					{
						return true;
					}
				}
			}
		}

		return false;
	}

	public static HashSet<ItemStack> getFilters(ItemStack itemStack)
	{
		HashSet<ItemStack> filterStacks = new HashSet<ItemStack>();

		NBTTagCompound nbt = NBTUtility.getNBTTagCompound(itemStack);
		NBTTagList tagList = nbt.getTagList("Items", 0);

		for (int i = 0; i < tagList.tagCount(); ++i)
		{
			NBTTagCompound var4 = (NBTTagCompound) tagList.getCompoundTagAt(i);
			filterStacks.add(ItemStack.loadItemStackFromNBT(var4));
		}

		return filterStacks;
	}

	public static List<ItemStack> getFilterList(ItemStack itemStack)
	{
		List<ItemStack> filterStacks = new ArrayList<ItemStack>();

		NBTTagCompound nbt = NBTUtility.getNBTTagCompound(itemStack);
		NBTTagList tagList = nbt.getTagList("Items", 0);

		for (int i = 0; i < tagList.tagCount(); ++i)
		{
			NBTTagCompound var4 = (NBTTagCompound) tagList.getCompoundTagAt(i);
			filterStacks.add(ItemStack.loadItemStackFromNBT(var4));
		}

		return filterStacks;
	}
}

File: src/main/scala/resonantinduction/archaic/blocks/ItemImprint.scala
package resonantinduction.archaic.blocks

import java.util
import java.util.{ArrayList, List}

import net.minecraft.entity.player.EntityPlayer
import net.minecraft.entity.{Entity, EntityList, EntityLivingBase, IProjectile}
import net.minecraft.item.{Item, ItemStack}
import net.minecraft.nbt.{NBTTagCompound, NBTTagList}
import resonant.lib.utility.LanguageUtility
import resonant.lib.utility.nbt.NBTUtility
import resonant.lib.wrapper.WrapList._
import resonantinduction.core.{Reference, RICreativeTab}

import scala.collection.JavaConversions._

object ItemImprint
{
    /**
     * Saves the list of items to filter out inside.
     */
    def setFilters(itemStack: ItemStack, filterStacks: java.util.List[ItemStack])
    {
        if (itemStack.getTagCompound == null)
        {
            itemStack.setTagCompound(new NBTTagCompound)
        }
        val nbt: NBTTagList = new NBTTagList
        import scala.collection.JavaConversions._
        for (filterStack <- filterStacks)
        {
            val newCompound: NBTTagCompound = new NBTTagCompound
            filterStack.writeToNBT(newCompound)
            nbt.appendTag(newCompound)
        }
        itemStack.getTagCompound.setTag("Items", nbt)
    }

    def isFiltering(filter: ItemStack, itemStack: ItemStack): Boolean =
    {
        if (filter != null && itemStack != null)
        {
            val checkStacks: List[ItemStack] = getFilters(filter)
            if (checkStacks != null)
            {
                import scala.collection.JavaConversions._
                for (stack <- checkStacks)
                {
                    if (stack.isItemEqual(itemStack))
                    {
                        return true
                    }
                }
            }
        }
        return false
    }

    def getFilters(itemStack: ItemStack): List[ItemStack] =
    {
        val filterStacks: List[ItemStack] = new util.LinkedList[ItemStack]
        val nbt: NBTTagCompound = NBTUtility.getNBTTagCompound(itemStack)
        val tagList: NBTTagList = nbt.getTagList("Items", 0)

        for (i <- 0 to tagList.tagCount)
        {
            val var4: NBTTagCompound = tagList.getCompoundTagAt(i)
            filterStacks.add(ItemStack.loadItemStackFromNBT(var4))
        }

        return filterStacks
    }

    def getFilterList(itemStack: ItemStack): List[ItemStack] =
    {
        val filterStacks: List[ItemStack] = new ArrayList[ItemStack]
        val nbt: NBTTagCompound = NBTUtility.getNBTTagCompound(itemStack)
        val tagList: NBTTagList = nbt.getTagList("Items", 0)

        for (i <- 0 to tagList.tagCount)
        {
            val var4: NBTTagCompound = tagList.getCompoundTagAt(i)
            filterStacks.add(ItemStack.loadItemStackFromNBT(var4))

        }
        return filterStacks
    }
}

class ItemImprint extends Item
{
    //Constructor
    setUnlocalizedName(Reference.prefix + "imprint")
    setTextureName(Reference.prefix + "imprint")
    setCreativeTab(RICreativeTab)
    setHasSubtypes(true)
    setMaxStackSize(1)

    override def onLeftClickEntity(stack: ItemStack, player: EntityPlayer, entity: Entity): Boolean =
    {
        if (entity != null && !(entity.isInstanceOf[IProjectile]) && !(entity.isInstanceOf[EntityPlayer]))
        {
            val stringName: String = EntityList.getEntityString(entity)
            return true
        }
        return false
    }

    def itemInteractionForEntity(par1ItemStack: ItemStack, par2EntityLiving: EntityLivingBase): Boolean =
    {
        return false
    }

    override def addInformation(itemStack: ItemStack, par2EntityPlayer: EntityPlayer, list: java.util.List[_], par4: Boolean)
    {
        val filterItems: List[ItemStack] = ItemImprint.getFilters(itemStack)

        if (filterItems.size > 0)
        {
            for (filterItem <- filterItems)
            {
              if(filterItem != null)
                list.add(filterItem.getDisplayName)
            }
        }
        else
        {
            list.add(LanguageUtility.getLocal("tooltip.noImprint"))
        }
    }
}